src/main/ruby/core/io.rb

```diff
@@ -272,12 +272,12 @@ def unseek!(io)
     ##
     # Returns +count+ bytes from the +start+ of the buffer as a new String.
     # If +count+ is +nil+, returns all available bytes in the buffer.
-    def shift(count=nil)
+    def shift(count=nil, encoding=Encoding::ASCII_8BIT)
       Truffle.synchronize(self) do
         total = size
         total = count if count and count < total
 
-        str = String.from_bytearray @storage, @start, total
+        str = String.from_bytearray @storage, @start, total, encoding
```

Comment:
Is there a use-case for a non-binary encoding?
I guess `shift` here is only called with 1 argument, but maybe another optimization can take advantage of it?

Reply:
I have a second set of changes where I create a UTF-8 string directly from the byte array. But it was starting to be a lot of changes for a single pull request, so I decided to split them.